cpp/ql/lib/semmle/code/cpp/security/InvalidPointerDereference/AllocationToInvalidPointer.qll

```diff
@@ -96,73 +96,111 @@ predicate hasSize(HeuristicAllocationExpr alloc, DataFlow::Node n, int state) {
-      state = getAFlowStateForNode(nRight) and
-      k <= state
+    flowTo(right) and
+    SizeBarrierConfig::isSink(DataFlow::operandNode(left), right, g, k, edge)
+  }
+
+  /**
+   * Gets an instruction `instr` that is guarded by a check such as `instr <= left + delta` where
+   * `left <= _ + k` and `left` is the "small side" of of a relational comparison that checks
+   * whether `left <= size` where `size` is the size of an allocation.
+   */
+  Instruction getABarrierInstruction0(int delta, int k) {
+    exists(IRGuardCondition g, ValueNumber value, Operand left, boolean edge, DataFlow::Node right |
+      // We know:
+      // 1. result <= value + delta (by `bounded`)
+      // 2. value <= right + k (by `operandGuardChecks`).
+      // So:
+      // result <= value + delta (by 1.)
+      //        <= right + k + delta (by 2.)
+      left = value.getAUse() and
+      operandGuardChecks(pragma[only_bind_into](g), pragma[only_bind_into](left), right,
+        pragma[only_bind_into](k), pragma[only_bind_into](edge)) and
+      bounded(result, value.getAnInstruction(), delta) and
+      g.controls(result.getBlock(), edge) and
+      k < getASizeAddend(right)
     )
   }
 
   /**
    * Gets an instruction that is guarded by a guard condition which ensures that
    * the value of the instruction is upper-bounded by size of some allocation.
    */
+  bindingset[state]
+  pragma[inline_late]
   Instruction getABarrierInstruction(int state) {
-    exists(IRGuardCondition g, ValueNumber value, Operand use, boolean edge |
-      use = value.getAUse() and
-      operandGuardChecks(pragma[only_bind_into](g), pragma[only_bind_into](use), _,
-        pragma[only_bind_into](state), pragma[only_bind_into](edge)) and
-      result = value.getAnInstruction() and
-      g.controls(result.getBlock(), edge)
+    exists(int delta, int k |
+      state > k + delta and
+      // result <= "size of allocation" + delta + k
+      //        < "size of allocation" + state
+      result = getABarrierInstruction0(delta, k)
     )
   }
 
+  private module ValidForStateConfig implements DataFlow::ConfigSig {
+    predicate isSource(DataFlow::Node source) { hasSize(_, source, _) }
+
+    predicate isSink(DataFlow::Node sink) { isSink(sink, _, _) }
+
+    additional predicate isSink(DataFlow::Node sink, int delta, int k) {
+      sink.asOperand() = SizeBarrier::getABarrierInstruction0(delta, k).getAUse()
+    }
+  }
+
+  private module ValidForStateFlow = DataFlow::Global<ValidForStateConfig>;
+
   /**
    * Gets a `DataFlow::Node` that is guarded by a guard condition which ensures that
    * the value of the node is upper-bounded by size of some allocation.
    */
   DataFlow::Node getABarrierNode(int state) {
-    result.asOperand() = getABarrierInstruction(state).getAUse()
+    exists(DataFlow::Node source, int delta, int k |
+      ValidForStateFlow::flow(source, result) and
+      hasSize(_, source, state) and
```

Comment:
For your earlier comments I got that `state` was coming from something related directly to the PAI. Here, however, it seems to be coming from the allocation?

Reply:
On second thought, I think I'm reading this incorrectly, but some clarification is welcome.

Reply:
That's a good point. Note that we have two kinds of barriers:
- `getABarrierNode` (i.e., this predicate) is used as a barrier on the second projection of the product-flow configuration. The `state` parameter here is the size addend that's added to the size of the allocation (i.e., `2` in `new int[size + 2]`).
- `getABarrierInstruction` (i.e., the predicate just above this one) is used as a barrier in `pointerAddInstructionHasBounds0` to check that:
  1. there's no guard that properly bounds the pai's right operand
  2. there's no guard that properly bounds the instruction `i` which satiesfies `i <= pai.getRight() + delta`.

The `state` parameter on `getABarrierInstruction` is the `delta` from point ii.

Reply:
Makes sense.